File: sdk/js/test/huggingFaceCatalog.test.ts
import { describe, it } from 'mocha';
import { expect } from 'chai';
import { getTestManager } from './testUtils.js';

const HF_URL = 'https://huggingface.co/onnxruntime/Phi-3-mini-4k-instruct-onnx/tree/main/cpu_and_mobile/cpu-int4-rtn-block-32-acc-level-4';

describe('HuggingFace Catalog Tests', () => {
    it('should create huggingface catalog', async function() {
        const manager = getTestManager();
        const hfCatalog = await manager.addCatalog('https://huggingface.co');
        expect(hfCatalog.name).to.equal('HuggingFace');
    });

    it('should reject non-huggingface url', async function() {
        const manager = getTestManager();
        try {
            await manager.addCatalog('https://example.com');
            expect.fail('Should have thrown an error for non-HuggingFace URL');
        } catch (error) {
            expect(error).to.be.instanceOf(Error);
            expect((error as Error).message).to.include('Unsupported catalog URL');
        }
    });

    it('should register model', async function() {
        this.timeout(30000);
        const manager = getTestManager();
        const hfCatalog = await manager.addCatalog('https://huggingface.co');

        const model = await hfCatalog.registerModel(HF_URL);
        expect(model.alias).to.be.a('string');
        expect(model.alias.length).to.be.greaterThan(0);
        expect(model.id).to.be.a('string');
        expect(model.id.length).to.be.greaterThan(0);
    });

    it('should find registered model by identifier', async function() {
        this.timeout(30000);
        const manager = getTestManager();
        const hfCatalog = await manager.addCatalog('https://huggingface.co');

        await hfCatalog.registerModel(HF_URL);

        const found = await hfCatalog.getModel(HF_URL);
        expect(found).to.not.be.undefined;
    });

    it('should register then download model', async function() {
        this.timeout(600000);
        const manager = getTestManager();
        const hfCatalog = await manager.addCatalog('https://huggingface.co');

        const model = await hfCatalog.registerModel(HF_URL);
        expect(model.alias.length).to.be.greaterThan(0);

        // Now download the ONNX files
        await model.download();
    });

    it('should reject registration of plain alias', async function() {
        const manager = getTestManager();
        const hfCatalog = await manager.addCatalog('https://huggingface.co');

        try {
            await hfCatalog.registerModel('phi-3-mini');
            expect.fail('Should have thrown an error for plain alias');
        } catch (error) {
            expect(error).to.be.instanceOf(Error);
            expect((error as Error).message).to.include('not a valid HuggingFace URL');
        }
    });

    it('should list registered models', async function() {
        this.timeout(30000);
        const manager = getTestManager();
        const hfCatalog = await manager.addCatalog('https://huggingface.co');

        await hfCatalog.registerModel(HF_URL);

        const models = await hfCatalog.getModels();
        expect(models).to.be.an('array');
        expect(models.length).to.be.greaterThan(0);
    });
});

File: sdk/rust/tests/integration/huggingface_catalog_test.rs
use super::common;

const HF_URL: &str = "https://huggingface.co/onnxruntime/Phi-3-mini-4k-instruct-onnx/tree/main/cpu_and_mobile/cpu-int4-rtn-block-32-acc-level-4";

#[tokio::test]
async fn should_create_huggingface_catalog() {
    let manager = common::get_test_manager();
    let hf_catalog = manager
        .add_catalog("https://huggingface.co", None)
        .await
        .expect("add_catalog failed");
    assert_eq!(hf_catalog.name(), "HuggingFace");
}

#[tokio::test]
async fn should_reject_non_huggingface_url() {
    let manager = common::get_test_manager();
    let result = manager.add_catalog("https://example.com", None).await;
    assert!(result.is_err(), "Non-HuggingFace URL should be rejected");
}

#[tokio::test]
async fn should_register_model() {
    let manager = common::get_test_manager();
    let hf_catalog = manager
        .add_catalog("https://huggingface.co", None)
        .await
        .expect("add_catalog failed");

    let model = hf_catalog
        .register_model(HF_URL)
        .await
        .expect("register_model failed");

    assert!(!model.alias().is_empty(), "Model alias should be non-empty");
    assert!(!model.id().is_empty(), "Model id should be non-empty");
}

#[tokio::test]
async fn should_find_registered_model_by_identifier() {
    let manager = common::get_test_manager();
    let hf_catalog = manager
        .add_catalog("https://huggingface.co", None)
        .await
        .expect("add_catalog failed");

    let _model = hf_catalog
        .register_model(HF_URL)
        .await
        .expect("register_model failed");

    let found = hf_catalog
        .get_model(HF_URL)
        .await
        .expect("get_model failed");

    assert!(!found.alias().is_empty(), "Should find model by HuggingFace URL");
}

#[tokio::test]
async fn should_register_then_download_model() {
    let manager = common::get_test_manager();
    let hf_catalog = manager
        .add_catalog("https://huggingface.co", None)
        .await
        .expect("add_catalog failed");

    let registered = hf_catalog
        .register_model(HF_URL)
        .await
        .expect("register_model failed");

    assert!(!registered.alias().is_empty());

    // Now download the ONNX files
    registered
        .download::<fn(&str)>(None)
        .await
        .expect("download failed");
}

#[tokio::test]
async fn should_reject_registration_of_plain_alias() {
    let manager = common::get_test_manager();
    let hf_catalog = manager
        .add_catalog("https://huggingface.co", None)
        .await
        .expect("add_catalog failed");

    let result = hf_catalog.register_model("phi-3-mini").await;
    assert!(result.is_err(), "Plain alias should be rejected");
}

#[tokio::test]
async fn should_list_registered_models() {
    let manager = common::get_test_manager();
    let hf_catalog = manager
        .add_catalog("https://huggingface.co", None)
        .await
        .expect("add_catalog failed");

    let _model = hf_catalog
        .register_model(HF_URL)
        .await
        .expect("register_model failed");

    let models = hf_catalog
        .get_models()
        .await
        .expect("get_models failed");

    assert!(
        !models.is_empty(),
        "Should have at least one registered model"
    );
}
